src/trans/gpu/CMakeLists.txt

```diff
@@ -6,115 +6,204 @@
+  PUBLIC_LIBS          fiat ectrans_common
+  PRIVATE_LIBS         ${ECTRANS_GPU_HIP_LIBRARIES}
+                       $<${HAVE_ACC}:OpenACC::OpenACC_Fortran>
+                       $<${HAVE_CUTLASS}:nvidia::cutlass::cutlass>
+  PRIVATE_DEFINITIONS  ${GPU_RUNTIME}GPU ${GPU_OFFLOAD}GPU
+                       $<${HAVE_CUTLASS}:USE_CUTLASS>
+                       $<${HAVE_CUTLASS_3XTF32}:USE_CUTLASS_3XTF32>
+                       $<${HAVE_GPU_GRAPHS_GEMM}:USE_GRAPHS_GEMM>
+)
+
+ectrans_target_fortran_module_directory(
+  TARGET            ectrans_gpu_common
+  MODULE_DIRECTORY  ${PROJECT_BINARY_DIR}/module/ectrans
+  INSTALL_DIRECTORY module/ectrans
+)
+
+
+
+function(generate_backend_sources)
+  set (options)
+  set (oneValueArgs BACKEND DESTINATION OUTPUT)
+  set (multiValueArgs)
+
+  cmake_parse_arguments(_PAR "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+  set(backend ${_PAR_BACKEND})
+  set(destination ${_PAR_DESTINATION})
+  file(MAKE_DIRECTORY ${destination}/algor)
+  file(MAKE_DIRECTORY ${destination}/internal)
+  file(MAKE_DIRECTORY ${destination}/external)
+
+  ecbuild_list_add_pattern( LIST files
+    GLOB
+          internal/*
+          external/*
+    QUIET
+  )
+  list( APPEND files
+          algor/seefmm_mix.F90
+          algor/wts500_mod.F90
+  )
+  ecbuild_list_exclude_pattern( LIST files REGEX
+          parkind_ectrans.F90
+          tpm_stats.F90
+          tpm_hicfft.F90
+  )
+
+  set(outfiles)
+  foreach(file_i ${files})
+    get_filename_component(outfile_name    ${file_i} NAME)
+    get_filename_component(outfile_name_we ${file_i} NAME_WE)
+    get_filename_component(outfile_ext     ${file_i} EXT)
+    get_filename_component(outfile_dir     ${file_i} DIRECTORY)
+    set(outfile "${destination}/${file_i}")
+    ecbuild_debug("Generate ${outfile}")
+    generate_file(BACKEND ${backend} INPUT ${CMAKE_CURRENT_SOURCE_DIR}/${file_i} OUTPUT ${outfile})
+    list(APPEND outfiles ${outfile})
+  endforeach(file_i)
+  set(${_PAR_OUTPUT} ${outfiles} PARENT_SCOPE)
+endfunction(generate_backend_sources)
+
+
+set( BUILD_INTERFACE_INCLUDE_DIR ${CMAKE_BINARY_DIR}/include/ectrans )
+
 foreach( prec dp sp )
   if( HAVE_${prec} )
-        set( GPU_LIBRARY_TYPE SHARED )
-        if( HAVE_GPU_STATIC )
-            set( GPU_LIBRARY_TYPE STATIC )
+
+        set(GENERATED_SOURCE_DIR ${CMAKE_CURRENT_BINARY_DIR}/generated/ectrans_gpu_${prec})
+
+        generate_backend_includes(BACKEND gpu_${prec} TARGET ectrans_gpu_${prec}_includes DESTINATION ${BUILD_INTERFACE_INCLUDE_DIR} INCLUDE_DIRECTORY ${PROJECT_SOURCE_DIR}/src/trans/include )
+        generate_backend_sources( BACKEND gpu_${prec} OUTPUT ectrans_gpu_${prec}_src  DESTINATION ${GENERATED_SOURCE_DIR})
+
+        ## Apply workarounds for some known compilers
+
+        if(CMAKE_Fortran_COMPILER_ID MATCHES "NVHPC")
+
+            # Compile setup_trans with pinned memory to improve data movement performance.
+            ectrans_add_compile_options(
```

Comment:
make sure to properly apply https://github.com/ecmwf-ifs/ectrans/commit/51cd73c0e959f3ead89a2033fa6f3553609e885a

Reply:
I will rebase and take extra care.

Reply:
Done